design/thin-edge-actors-design.md

```diff
@@ -553,23 +583,273 @@ impl From<Msg2> for Msg {
+
+```rust
+pub trait Actor {
+  
+    /// Type of output messages this actor produces
+    type Output: Message;
+  
+    /// Type of the peers this actor interacts with
+    type Peers: From<Recipient<Self::Output>>;
+    
+    // ...
+}
+```
+
+Note that the actor `Output` type can be built in a similar fashion as its `Input` type,
+using the `fan_in_message_type!` macro.
+For instance, the expression `fan_in_message_type!(C8yMapperOutput[SoftwareRequest,OperationRequest]);` expends to:
+
+```rust
+#[derive(Clone, Debug)]
+enum C8yMapperOutput {
+  SoftwareRequest(SoftwareRequest),
+  OperationRequest(OperationRequest),
+}
+
+impl From<SoftwareRequest> for C8yMapperOutput {
+    fn from(m: SoftwareRequest) -> C8yMapperOutput {
+      C8yMapperOutput::SoftwareRequest(m)
+    }
+}
+
+impl From<OperationRequest> for C8yMapperOutput {
+    fn from(m: OperationRequest) -> C8yMapperOutput {
+      C8yMapperOutput::OperationRequest(m)
+    }
+}
+```
+
+Similarly, the `fan_out_peer_type!` macro can be used to create the boilerplate code required
+to build a struct of peers from a recipient that gather all the individual messages.
+
+* TODO: describe the `fan_out_peer_type!` macro that creates a struct of peers each with a specific type.
+
+Note also that each individual peer doesn't have to be represented by a `Recipient<T>` of the appropriate type.
+It can be a smart handler that hides the message sending API behind regular rust methods.
+
+```rust
+struct C8yMapperPeers {
+  sw_manager: SWManager,
+  op_schedule: OpScheduler,
+}
+
+impl From<Recipient<C8yMapperOutput>> for C8yMapperPeers {
+  fn from(recipient: Recipient<C8yMapperOutput>) -> Self {
+    C8yMapperPeers {
+      sw_manager: SWManager::from(recipient),
+      op_schedule: OpScheduler::from(recipient),
+    }
+  }
+}
+
+struct SWManager {
+  sender: Recipient<SoftwareRequest>
+}
+
+impl SWManager {
+  pub async fn do_this(&mut self, arg: ThisArg) -> Result<(), ChannelError> {
+    let req : SoftwareRequest = arg.into();
+    self.sender.send(req).await
+  }
+
+  pub async fn do_that(&mut self, arg: ThatArg) -> Result<(), ChannelError> {
+    let req : SoftwareRequest = arg.into();
+    self.sender.send(req).await
+  }
+}
+
+struct OpScheduler {
+  sender: Recipient<OperationRequest>
+}
+
+impl OpScheduler {
+  pub async fn say(&mut self, arg: &str) -> Result<(), ChannelError> {
+    let req : OperationRequest = arg.into();
+    self.sender.send(req).await
+  }
+}
+```
 
-### Runtime messages
+To be discussed:
+* Can these smart handlers be generated by a macro?
+
+#### Abstracting channels from peers
+
+#### Actor with no inputs or outputs
+
+An actor can have no input messages and only acts as a source of output messages.
+Similarly, an actor can have not output messages and be a sink of input messages.
+Notable examples are respectively a source of measurements and a message logger.
```

Comment:
Having a single centralised logging Actor could be a bit complex to maintain when the target of different actors could be separate log files on the disk. Sounds like an unnecessary complexity for now. Leaving the logging to individual actors themselves would be simpler, I feel.

Reply:
> Having a single centralised logging Actor could be a bit complex to maintain when the target of different actors could be separate log files on the disk. Sounds like an unnecessary complexity for now. Leaving the logging to individual actors themselves would be simpler, I feel.

I agree. That said, at this design stage, it's important to understand what would be necessary to implement such a logger.